### PR TITLE
o/servicestate: unlock state before calling wrappers in doServiceControl

### PR DESCRIPTION
**Pull request by stolowski**
Keeping the lock for the duration of doServiceControl causes massive slowdown/timeout e.g. on lxd start. On startup lxd calls `snap model ...` and it takes around 4 minutes to handle (client is blocked for that long when handling snap start):

```
Nov 16 15:48:40 nov161534-153926 snapd[22157]: daemon.go:317: DEBUG: pid=23342;uid=0;socket=/run/snapd.socket; POST /v2/apps 9.26145ms 202
Nov 16 15:48:40 nov161534-153926 snapd[22157]: taskrunner.go:439: DEBUG: Running task 56 on Do: Run service command "start" for services ["activate" "daemon"] of snap "lxd"
Nov 16 15:52:40 nov161534-153926 snapd[22157]: daemon.go:317: DEBUG: pid=23432;uid=0;socket=/run/snapd.socket; GET /v2/model/serial 4m0.032733113s 200
Nov 16 15:52:40 nov161534-153926 snapd[22157]: daemon.go:317: DEBUG: pid=23432;uid=0;socket=/run/snapd.socket; GET /v2/model 2m0.03386201s 200
```

### CODE REVIEW

File: overlord/servicestate/service_control.go [hunk 1]
```diff
@@ -87,15 +87,19 @@ func (m *ServiceManager) doServiceControl(t *state.Task, _ *tomb.Tomb) error {
...
 	case "stop":
 		disable := sc.ActionModifier == "disable"
 		flags := &wrappers.StopServicesFlags{
 			Disable: disable,
 		}
-		if err := wrappers.StopServices(services, flags, snap.StopReasonOther, meter, perfTimings); err != nil {
+		st.Unlock()
+		err := wrappers.StopServices(services, flags, snap.StopReasonOther, meter, perfTimings)
+		st.Lock()
+		if err != nil {
```
Comment (pedronis) [inline overlord/servicestate/service_control.go:102]: given that we might change and write snapst back,  the safest is probably to read it again after we have reacquired the lock

File: overlord/servicestate/service_control.go [hunk 2]
```diff
@@ -116,7 +120,10 @@ func (m *ServiceManager) doServiceControl(t *state.Task, _ *tomb.Tomb) error {
 		flags := &wrappers.StartServicesFlags{
 			Enable: enable,
 		}
-		if err := wrappers.StartServices(startupOrdered, nil, flags, meter, perfTimings); err != nil {
+		st.Unlock()
+		err = wrappers.StartServices(startupOrdered, nil, flags, meter, perfTimings)
+		st.Lock()
```
Comment (pedronis) [inline overlord/servicestate/service_control.go:125]: same

**Review by pedronis — CHANGES_REQUESTED** [review]
some considerations

File: overlord/servicestate/service_control.go (hunk 1 above)
Comment (stolowski) [reply overlord/servicestate/service_control.go:102]: Updated.

**Review by pedronis — APPROVED** [review]
thanks

**Review by mvo5 — APPROVED** [review]
Thanks for finding this bug!

**Pull request merged**